### PR TITLE
feature: misc async operations

### PR DESCRIPTION
**Pull request by rgrinberg**
This moves over the remaining operations in the engine into background
threads. They aren't really related to each other in any way, so I'm not
sure what to call the config to toggle them.

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

### CODE REVIEW

**Review by snowleopard — COMMENTED** [review]
How about calling the toggle `background_file_system_operations_in_rule_execution`?

**Comment by rgrinberg** [comment]
> How about calling the toggle `background_file_system_operations_in_rule_execution`?

Done

File: src/dune_engine/targets.mli [hunk 1]
```diff
@@ -114,3 +115,7 @@ module Produced : sig
 
   val to_dyn : _ t -> Dyn.t
 end
+
+(** will run in a background thread if
+    [background_file_system_operations_in_rule_execution] is set *)
+val maybe_async : (unit -> 'a) -> 'a Fiber.t
```
Comment (snowleopard) [inline src/dune_engine/targets.mli:121]: Seems like an odd place to put this function.

**Review by snowleopard — APPROVED** [review]
Looks good!

**Pull request merged**